crates/unftp-sbe-fs/src/cap_fs.rs

```diff
@@ -0,0 +1,106 @@
+//! A capabilities-friendly workalike of tokio::fs
+// Most of these functions are copied almost verbatim from tokio::fs, but with the std parts
+// replaced by cap_std.
+
+use std::{io, path::Path, sync::Arc};
+
+use tokio::{sync::mpsc, task::spawn_blocking};
+use tokio_stream::wrappers::ReceiverStream;
+
+/// Exact copy of tokio::fs::asyncify
+async fn asyncify<F, T>(f: F) -> io::Result<T>
+where
+    F: FnOnce() -> io::Result<T> + Send + 'static,
+    T: Send + 'static,
+{
+    match spawn_blocking(f).await {
+        Ok(res) => res,
+        Err(_) => Err(io::Error::new(io::ErrorKind::Other, "background task failed")),
+    }
+}
+
+/// Create a new directory somewhere under this one
+pub async fn create_dir<P: AsRef<Path>>(root: Arc<cap_std::fs::Dir>, path: P) -> io::Result<()> {
+    let path = path.as_ref().to_owned();
+    asyncify(move || root.create_dir(path)).await
+}
+
+pub async fn open<P: AsRef<Path>>(root: Arc<cap_std::fs::Dir>, path: P) -> io::Result<cap_std::fs::File> {
+    let path = path.as_ref().to_owned();
+    asyncify(move || root.open(path)).await
+}
+
+pub async fn open_with<P: AsRef<Path>>(root: Arc<cap_std::fs::Dir>, path: P, options: cap_std::fs::OpenOptions) -> io::Result<cap_std::fs::File> {
+    let path = path.as_ref().to_owned();
+    asyncify(move || root.open_with(path, &options)).await
+}
+
+/// Returns a stream over the entries within a directory.
+///
+/// This is a capabilties-based, async version of [`std::fs::read_dir`](std::fs::read_dir)
+///
+/// This operation is implemented by running the equivalent blocking
+/// operation on a separate thread pool using [`spawn_blocking`].
```

Comment:
I'm starting to think this should be its own storage back-end.

Reply:
Remember, Tokio itself also uses `spawn_blocking` for reading a directory.